### PR TITLE
Updated udev rules

### PR DESCRIPTION
**Pull request by aschneid1**
This PR adds additional udev rules and a script to support disk identification for PD NVMe disks.  

It requires the nvme-cli tool (updated in package configurations) and a specific vendor extension to the NVMe device/controller.  The disk name will be present in this field (along with some other metadata).

It was tested in the following configurations:
    - Devcluster with multiple NVMe disks attached (symlinks created including links for specific partitions)
    - Staging & Devcluster with missing nvme-cli tool (boot successful but no symlink, which is expected)
    - Staging with missing vendor extension (boot successful and no symlink, which is expected)

### CODE REVIEW

File: src/lib/udev/google_nvme_id [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+#######################################
+function identify_pd_disk() {
+  local dev_path="${1}"
+  local dev_name
+  dev_name="$(get_namespace_device_name "${dev_path}")"
+  if [[ $? != 0 ]]; then
+    return 255
+  fi
+
+  ID_NAMESPACE_INDEX="${namespace_index}"
```
Comment (hopkiw) [inline src/lib/udev/google_nvme_id:202]: where does $namespace_index come from? where is $ID_NAMESPACE_INDEX used?

File: src/lib/udev/google_nvme_id [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+#######################################
+# Ensures a given command/path exists
+# Arguments:
+#   The path for the command
+# Returns:
+#   0 if the command is found/executable and 1 on error
+#######################################
+function confirm_command() {
+  local command_to_check="${1}"
+  command -v "${command_to_check}" > /dev/null 2>&1
```
Comment (hopkiw) [inline src/lib/udev/google_nvme_id:217]: this single line can be used in place of this function

File: src/lib/udev/google_nvme_id [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+# Validates that a given path corresponds to a valid NVMe device
+# Globals:
+#   Uses nvme_cli_bin
+# Arguments:
+#   The device path for the disk
+# Returns:
+#   The return value of the call to the nvme id-ctrl command
+#######################################
+function validate_nvme_device() {
+  "${nvme_cli_bin}" id-ctrl "${1}" &>/dev/null
```
Comment (hopkiw) [inline src/lib/udev/google_nvme_id:160]: this single line can be used in place of this function

File: src/lib/udev/google_nvme_id [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+    err_msg="${err_msg} You may need to run with sudo or install nvme-cli."
+    err "${err_msg}"
+    return 1
+  fi
+
+  # Ensure the passed device is actually an NVMe device
+  validate_nvme_device "${device_path}"
+  if [[ $? != 0 ]]; then
+    local err_msg="Passed device was not an NVMe device.  (You may need to run"
+    err_msg="${err_msg} this script as root/with sudo)."
```
Comment (hopkiw) [inline src/lib/udev/google_nvme_id:269]: nit, style: line escapes are easier to read

File: src/lib/udev/google_nvme_id [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+  # Detect the type of attached nvme device
+  local device_type="$(detect_pd_or_local_ssd "${device_path}")"
+  local device_type_ret_value=$?
+  if [[ "${device_type}" == "${LOCAL_SSD}" ]]; then
+    err "This script only supports PD NVMe disks but given path was Local SSD"
+    return 1
+  elif [[ "${device_type_ret_value}" != 0 ]]; then
+    err "Unknown device type (neither a PD nor Local SSD NVMe device)"
+    return 1
+  fi
```
Comment (hopkiw) [inline src/lib/udev/google_nvme_id:283]: this is overcomplicated, imo. the detect_pd_or_local_ssd function checks for two hardcoded values, translating them into one of two different globals. then in use here, the two globals are checked against. so the globals were not used, and we could have just checked against the real command output here directly

try:
```
local device_type
if ! device_type=$("$nvme_cli_bin" id-ctrl "$device_path"); then
  err "Unknown device type [...]"
  return 1
fi
if [[ "$device_type" != "nvme_card-pd" ]]; then
  err "Unsupported device type [...]"
  return 1
fi
```

File: src/lib/udev/google_nvme_id [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+#   The device path for the disk
+# Returns:
+#   0 on success and 255 if an error occurrs
+#######################################
+function identify_pd_disk() {
+  local dev_path="${1}"
+  local dev_name
+  dev_name="$(get_namespace_device_name "${dev_path}")"
+  if [[ $? != 0 ]]; then
+    return 255
```
Comment (hopkiw) [inline src/lib/udev/google_nvme_id:199]: why 255? other functions return 1 on error

File: src/lib/udev/google_nvme_id (hunk 1 above)
Comment (hopkiw) [reply src/lib/udev/google_nvme_id:202]: also note, you are setting globals as a side effect of invoking the function, rather than using return values or outputs. that's hard to follow, if you can easily avoid it you should

File: src/lib/udev/google_nvme_id [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+    err "Unknown device type (neither a PD nor Local SSD NVMe device)"
+    return 1
+  fi
+
+  # Fill the global variables for the id command for the given disk type
+  identify_pd_disk "${device_path}"
+  identify_ret_value=$?
+
+  # Error messages will be printed closer to error, no need to reprint here
+  if [[ "${identify_ret_value}" != 0 ]]; then
```
Comment (hopkiw) [inline src/lib/udev/google_nvme_id:290]: nit: these lines can be combined if we aren't going to use identify_ret_value again

File: src/lib/udev/google_nvme_id [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+    return "${identify_ret_value}"
+  fi
+
+  # Gen symlinks or print out the globals set by the identify command
+  if [[ "${opt_gen_symlink}" == 'true' ]]; then
+    gen_symlink "${device_path}"
+  else
+    echo "ID_SERIAL_SHORT=${ID_SERIAL_SHORT}"
+    echo "ID_SERIAL=${ID_SERIAL}"
```
Comment (hopkiw) [inline src/lib/udev/google_nvme_id:299]: these are the primary outputs of the script in normal operation, right?

**Review by hopkiw — CHANGES_REQUESTED** [review]
can we see some testing data for what some of these values, such as the JSON output from nvme-cli looks like and what the resulting symlinks are?
do we intend to support pd-nvme for boot disks? by symlink?

/ok-to-test
/approve

(approve only enables the PR for review)

File: src/lib/udev/google_nvme_id (hunk 3 above)
Comment (aschneid1) [reply src/lib/udev/google_nvme_id]: Fixed in most recent commit.

File: src/lib/udev/google_nvme_id (hunk 6 above)
Comment (aschneid1) [reply src/lib/udev/google_nvme_id]: Fixed in most recent commit.

File: src/lib/udev/google_nvme_id (hunk 1 above)
Comment (aschneid1) [reply src/lib/udev/google_nvme_id]: Hold over from a previous version of the code.  W.r.t the globals: it makes sense to me to keep those as globals since they're the primary output of the script, and it would be hard and a bit kludgy to split them out into their own return values.  I've added a comment to their declaration clarifying where they're set and why.

File: src/lib/udev/google_nvme_id (hunk 2 above)
Comment (aschneid1) [reply src/lib/udev/google_nvme_id]: Done.

File: src/lib/udev/google_nvme_id (hunk 4 above)
Comment (aschneid1) [reply src/lib/udev/google_nvme_id]: Done.

File: src/lib/udev/google_nvme_id (hunk 5 above)
Comment (aschneid1) [reply src/lib/udev/google_nvme_id]: Updated to remove the globals/detect_pd_or_local_ssd function.  These globals were originally used in multiple places (when the script was also used for local SSD).

File: src/lib/udev/google_nvme_id (hunk 7 above)
Comment (aschneid1) [reply src/lib/udev/google_nvme_id]: Done.

File: src/lib/udev/google_nvme_id (hunk 8 above)
Comment (aschneid1) [reply src/lib/udev/google_nvme_id:239]: Yep.  They get consumed by the updated udev rules.

File: src/lib/udev/google_nvme_id [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+#   The path to the nvme namespace (/dev/nvme0n?)
+# Outputs:
+#   The device name parsed from the JSON in the vendor ext of the ns-id command.
+# Returns:
+#   0 if the device name for the namespace could be retrieved, 1 otherwise
+#######################################
+function get_namespace_device_name() {
+  local nvme_json
+  nvme_json="$("${nvme_cli_bin}" id-ns -b "${1}" | xxd -p -seek 384 -l 256 | xxd -p -r)"
+  if [[ $? != 0 ]]; then
```
[the remark this answers is not in the document]
Comment (aschneid1) [reply src/lib/udev/google_nvme_id]: Fixed.

File: src/lib/udev/google_nvme_id [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+# Arguments:
+#   The path to the nvme namespace (/dev/nvme0n?)
+# Outputs:
+#   The device name parsed from the JSON in the vendor ext of the ns-id command.
+# Returns:
+#   0 if the device name for the namespace could be retrieved, 1 otherwise
+#######################################
+function get_namespace_device_name() {
+  local nvme_json
+  nvme_json="$("${nvme_cli_bin}" id-ns -b "${1}" | xxd -p -seek 384 -l 256 | xxd -p -r)"
```
[the remark this answers is not in the document]
Comment (aschneid1) [reply src/lib/udev/google_nvme_id]: Fixed.

**Comment by aschneid1** [comment]
> can we see some testing data for what some of these values, such as the JSON output from nvme-cli looks like and what the resulting symlinks are?

The symlinks should be "google-<pd-device-name>", where pd-device-name corresponds to whatever the user chose for the device name (defaults to the guest's disk name).  The json output from nvme-cli is currently experiment guarded and has to be ran in a devcluster for testing, but looks like:

{"device_name":"my-pd-test-short-name","disk_type":"PERSISTENT"}

(more fields may be added later).

> do we intend to support pd-nvme for boot disks? 

Yes

> by symlink?

Presumably, yes.  Although the behavior for boot disks shouldn't differ AFAIK.

File: src/lib/udev/google_nvme_id [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+    return 1
+  fi
+
+  if [[ -z "$nvme_json" ]]; then
+    err "NVMe Vendor Extension disk information not present"
+    return 1
+  fi
+
+  local device_name
+  device_name="$(echo "$nvme_json" | sed -e 's/.*"device_name":"\([a-zA-Z0-9_-]\+\)".*/\1/g')"
```
Comment (hopkiw) [inline src/lib/udev/google_nvme_id:69]: i think this needs careful consideration. JSON is a structured format, and this is a simple string parser, and it is not a safe way to do the transformation you want. have you considered writing this script in another language? is a dependency on jq possible?

Comment (aschneid1) [reply src/lib/udev/google_nvme_id:69]: It could be re-written in python so long as we could guarantee a stable release version across our supported distros (i.e., if everything had a python3 it shouldn't be problematic, save for the call-out to the shell to invoke nvme-cli which has its own set of problems [not a python expert, so I'm not sure how frowned upon external invocations like that via os are]). 

I'm trying to limit the number of external dependencies required and am not familiar with jq.  Open to your advice/suggestions here, though.

Comment (hopkiw) [reply src/lib/udev/google_nvme_id:69]: well, we actually moved a lot of our software away from python bc of the difficulties in guaranteeing consistency, shipping for different versions, etc. 

i will see how widely available jq is or if there is an alternate suggestion. but json has multiple string representations and it's tough to use something like sed safely with it, especially in a future-proof way. here are three string representations of the same json object, which would have different parsed results using the proposed sed command:

```
{"device_name":"my-pd-test-short-name","disk_type":"PERSISTENT"}
```

```
{
  "device_name": "my-pd-test-short-name",
  "disk_type": "PERSISTENT"
}
```

```
{
  "disk_type":"PERSISTENT",
  "device_name":"my-pd-test-short-name"
}
```

Comment (aschneid1) [reply src/lib/udev/google_nvme_id:69]: Makes sense.  FWIW, if jq turns out to be viable, the replacement is pretty straightforward:

echo "$nvme_json" | jq -r .device_name

Comment (hopkiw) [reply src/lib/udev/google_nvme_id:69]: yes, jq also supports hard-to-sed JSON like objects where the key and value are separated by a newline. unfortunately it doesn't seem universally available on supported images. so for that reason, i propose we just try to make this as safe as possible by adding a grep to select the specific line, and allowing for possible whitespace between key and value. suggest amending to one of:
```
device_name="$(echo "$nvme_json" | grep device_name | sed -e 's/.*"device_name":[  \t]*"\([a-zA-Z0-9_-]\+\)".*/\1/')"
```

```
device_name="$(echo "$nvme_json" | grep device_name | cut -d: -f2 | tr -d ' ",')"
```

Comment (aschneid1) [reply src/lib/udev/google_nvme_id]: Updated to the first option and verified in a devcluster

File: packaging/google-compute-engine.spec [hunk 12]
```diff
@@ -30,6 +30,7 @@ Requires: dracut
 Requires: google-compute-engine-oslogin
 Requires: google-guest-agent
 Requires: rsyslog
+Requires: nvme-cli
```
Comment (hopkiw) [inline packaging/google-compute-engine.spec:33]: you need to add a line to the el6 spec in the %install section that removes the lib/udev/google_nvme_id script. otherwise  el6 RPM would contain the script but not the nvme-cli dependency.

**Review by hopkiw — CHANGES_REQUESTED** [review]

File: packaging/google-compute-engine.spec (hunk 12 above)
Comment (aschneid1) [reply packaging/google-compute-engine.spec:33]: How does this jive with the inclusion of google_nvme_id in the 65-* rules file that requires google_nvme_id?  Is it okay to have a rule that depends on a script that doesn't exist?  Would it be better to leave the script in but have it fail due to the lack of nvme-cli?

Comment (aschneid1) [reply packaging/google-compute-engine.spec:33]: Updated to remove nvme-cli as asked.

Comment (hopkiw) [reply packaging/google-compute-engine.spec:33]: good point. this change to 65- should only be on el7+, as discussed. however i think there's a broader fix here, we can probably just prune el6 altogether here since it's now EOL. will confirm this and update

Comment (hopkiw) [reply packaging/google-compute-engine.spec:33]: yes, we can prune all el6 support. i will do this in a separate PR
